### DISCUSSION TITLE
Integrate Arcus.Observability in MessagePumps for automatically logging request

### DISCUSSION THREAD

**Discussion by fgheysels** [arcus-azure/arcus.messaging#263] [Ideas]
I think it would be a good idea if we make sure that a MessagePump logs a request (using Arcus.Observability) whenever a message is retrieved / processed from the underlying servicebus.
Of course, this also implies that it must be possible for the user of Arcus.Messaging to define some configuration-options as in:  where can I find an ID in the message that I've received that must be used for correlation, etc...

Or, do you think this is the responsability of the MessageHandler ?  But in that case, it is the developer of the specific MessageHandler that has the responsability of making sure that a ServiceBus request is logged (which might mean repeated code).

**Comment by stijnmoreels** [comment]
I think this is already proposed here: #221 , but indeed, without yet configuration. Maybe that's then an improvement? I think, for now, the PoC already gives some guidance as to where this information is located. the `ServiceBusReceivedMessage` has already a correlation ID default property, too. But yeah, could/should be configurable later.

**Reply by stijnmoreels** [reply]
I do not think it's the responsibility of the message handler to provide information of the request tracking. The sole purpose of the message handlers is message processing (much like the remainder of middleware components when we track HTTP requests). I think the way we configure now the message pump (with options) is a good place to put this.

**Reply by fgheysels** [reply]
I did a quick scan over the existing issues in Messaging, but I must have overlooked it.
I'm currently looking at / studying the PoC :D

In there, I can see that the MessagePump indeed logs the request, but this is indeed currently missing in the Arcus implementation.

**Reply by stijnmoreels** [reply]
We can consider moving the request tracking to the message router as there is no message pump when using the functionality in Azure Functions. Also mentioned in https://github.com/arcus-azure/arcus.messaging/issues/221

**Reply by fgheysels** [reply]
I think the 'configuration' part is mandatory, as afaik you cannot rely on a default header where you could look for the dependencyId.
The component that creates and posts the message to the service-bus will most likely add an additional 'user property' to the message which contains the dependencyId (correlation-id) that must be used by the component that processes the message to correlate the request.

Right now, the Service-to-Service PoC logs the request, but without specifying the dependency-Id:

![image](https://user-images.githubusercontent.com/3605786/161227377-8f02151b-c14e-4efa-91be-856b21eade6d.png)
[source](https://github.com/arcus-azure/arcus-service-to-service-correlation-poc/blob/12dba574c89a4aa9ea3e192d5887f5749b298f2e/src/Arcus.POC.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs#L596)

I'd say that , in real life, there is some kind of 'agreement' between message producer and message consumer on where to find the Id that can be used for correlating dependencies & requests.   So, the message consumer will need to specify somewhere in his code where / how to retrieve that Id from the message.

**Reply by stijnmoreels** [reply]
Now the dependency ID is retrieved back in the `.GetCorrelationInfo` extension on the Azure Service Bus `Message`. `OperationParentId`.
Indeed, should all be configurable.

**Reply by fgheysels** [reply]
> We can consider moving the request tracking to the message router as there is no message pump when using the functionality in Azure Functions. Also mentioned in #221

I've been through some of the documentation and the code.   I wouldn't say we can consider to move this request tracking to the MessageRouter, I think it is a given to have this request tracking implemented in the MessageRouter :).
Should this be defined in the `MessageRouterOptions` type ?

OTOH, I think Azure Functions will already log the request ?

**Reply by stijnmoreels** [reply]
> I've been through some of the documentation and the code. I wouldn't say we can consider to move this request tracking to the MessageRouter, I think it is a given to have this request tracking implemented in the MessageRouter :).
Should this be defined in the MessageRouterOptions type ?

Hihi, yeah 😄 .
Yep, that is I think a good place for this.

> OTOH, I think Azure Functions will already log the request ?

Well, hmm, maybe, but if we custom configure the dependency ID in a custom defined HTTP header/Azure Service Bus user property, I don't know if the built-in request tracking will be able to be configured to pick that up?

**Reply by stijnmoreels** [reply]
Done in https://github.com/arcus-azure/arcus.messaging/pull/266
